File: apps/atlas/lib/track/track_recorder.dart
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

import 'package:atlas_location/atlas_location.dart';
import 'package:flutter/foundation.dart';

import '../location/location_service.dart';

enum TrackRecorderState { idle, recording }

final class TrackRecorder extends ChangeNotifier {
  TrackRecorder({required LocationService locationService})
      : _location = locationService;

  final LocationService _location;
  TrackRecorderState _state = TrackRecorderState.idle;
  AtlasTrackLog _log = AtlasTrackLog();
  AtlasLocationFix? _lastAppended;
  bool _listening = false;
  bool _disposed = false;

  TrackRecorderState get state => _state;
  bool get isRecording => _state == TrackRecorderState.recording;
  int get pointCount => _log.fixCount;
  List<AtlasLocationFix> get points => _log.fixes;

  void start() {
    if (_disposed || _state == TrackRecorderState.recording) return;
    _log = AtlasTrackLog();
    _lastAppended = null;
    _state = TrackRecorderState.recording;
    if (!_listening) {
      _location.addListener(_onLocation);
      _listening = true;
    }
    _ingestCurrent();
    notifyListeners();
  }

  List<AtlasLocationFix> stop() {
    if (_state == TrackRecorderState.idle) return const [];
    _detach();
    final fixes = _log.fixes;
    _log = AtlasTrackLog();
    _lastAppended = null;
    _state = TrackRecorderState.idle;
    notifyListeners();
    return fixes;
  }

  @override
  void dispose() {
    _disposed = true;
    _detach();
    super.dispose();
  }

  void _detach() {
    if (_listening) {
      _location.removeListener(_onLocation);
      _listening = false;
    }
  }

  void _onLocation() {
    if (_disposed || _state != TrackRecorderState.recording) return;
    if (_location.status != AtlasLocationStatus.valid) return;
    final fix = _location.latestFix;
    if (fix == null || fix == _lastAppended) return;
    _log.append(fix);
    _lastAppended = fix;
    notifyListeners();
  }

  void _ingestCurrent() {
    if (_location.status != AtlasLocationStatus.valid) return;
    final fix = _location.latestFix;
    if (fix == null) return;
    _log.append(fix);
    _lastAppended = fix;
  }
}

File: apps/atlas-android/app/src/main/java/com/sovereignatlas/atlas/track/TrackRecorder.kt
// ============================================================
// As Above, So Below. As Within, So Without.
// The Future Dictates the Past and the Past is Always Present.
// ============================================================

package com.sovereignatlas.atlas.track

import com.sovereignatlas.atlas.location.AtlasLocationFix
import com.sovereignatlas.atlas.location.AtlasLocationStatus
import com.sovereignatlas.atlas.location.LocationService

enum class TrackRecorderState { idle, recording }

class TrackRecorder(private val location: LocationService) {
    private var state = TrackRecorderState.idle
    private val log = ArrayList<AtlasLocationFix>()
    private var lastAppended: AtlasLocationFix? = null
    private var listening = false
    private var disposed = false
    private val listeners = ArrayList<() -> Unit>()

    fun addListener(listener: () -> Unit) {
        listeners.add(listener)
    }

    fun removeListener(listener: () -> Unit) {
        listeners.remove(listener)
    }

    fun state(): TrackRecorderState = state

    fun isRecording(): Boolean = state == TrackRecorderState.recording

    fun pointCount(): Int = log.size

    fun points(): List<AtlasLocationFix> = log.toList()

    fun start() {
        if (disposed || state == TrackRecorderState.recording) return
        log.clear()
        lastAppended = null
        state = TrackRecorderState.recording
        if (!listening) {
            location.addListener(::onLocation)
            listening = true
        }
        ingestCurrent()
        notifyListeners()
    }

    fun stop(): List<AtlasLocationFix> {
        if (state == TrackRecorderState.idle) return emptyList()
        detach()
        val fixes = log.toList()
        log.clear()
        lastAppended = null
        state = TrackRecorderState.idle
        notifyListeners()
        return fixes
    }

    fun dispose() {
        disposed = true
        detach()
    }

    private fun detach() {
        if (listening) {
            location.removeListener(::onLocation)
            listening = false
        }
    }

    private fun onLocation() {
        if (disposed || state != TrackRecorderState.recording) return
        ingestCurrent()
        notifyListeners()
    }

    private fun ingestCurrent() {
        if (location.status() != AtlasLocationStatus.valid) return
        val fix = location.latestFixOrNull() ?: return
        if (fix === lastAppended) return
        log.add(fix)
        lastAppended = fix
    }

    private fun notifyListeners() {
        for (listener in listeners.toList()) {
            listener()
        }
    }
}
